security-force-monitor/sfm-cms
# Migrate entity search to Haystack

## Overview

This PR:
- Replaces the custom Solr search back end that powers entity search with a conventional implementation of Haystack. N.b., it does not migrate the logic to populate and query the composition documents to form the org charts from #416.
- Updates the startup instructions in the README and makes some small adjustments to the Docker setup.

Connects #674.

### Demo

![Screen Shot 2021-01-26 at 1 59 35 PM](https://user-images.githubusercontent.com/12176173/105898317-e6884200-5fde-11eb-90cf-338968fa364e.png)
![Screen Shot 2021-01-26 at 2 00 15 PM](https://user-images.githubusercontent.com/12176173/105898324-e7b96f00-5fde-11eb-8a0e-ea5649e17330.png)
![Screen Shot 2021-01-26 at 2 00 39 PM](https://user-images.githubusercontent.com/12176173/105898328-e8520580-5fde-11eb-8849-14b1e104cce0.png)

## Notes

There are a few classes of changes here:

- Adds `${ENTITY_MODULE}.search_indexes` for each search entity: person, unit, violation, and source.
- Updates `schema.xml` to include Haystack fields. N.b., the template into which these fields are generated lives in `templates/search_configs/schema.xml` and is a lightly edited version of the original schema.
- Replaces the custom search view with a Haystack `FacetedSearchView` and adjusts the search templates to work with the new context.
- Renames `make_search_index` to `update_composition_index` and removes code to add documents related to the base search entities.
- Updates the encrypted configs to include `HAYSTACK_*` variables (as exemplified in `local_settings.example.py`).
- Updates the tests to remove tests for index updates on model changes. (We'll be using Haystack's `RealtimeSignalProcessor` for this behavior, i.e., we do not need to test third-party code.)

## Testing Instructions

* If you do not have a local instance, follow the updated README instructions to retrieve a dump of the staging database and load it into your database. 
* Build the entity search index: `docker-compose run --rm app python manage.py update_index -v 3`
  * Note that this will take about half an hour.
* Build the composition search index: `docker-compose run --rm app python manage.py make_search_index`
* Create a superuser and log in so you can access source search: `docker-compose run --rm app python manage.py createsuperuser`
* Navigate to your local homepage and begin to execute a variety of searches. Confirm that load times are reasonable, and that the results generally match those on https://back.securityforcemonitor.org.
* Continue to make test searches from the search results page. Change sorting and paging parameters, and confirm the expected results resemble staging.
* Use the personnel search to find a commander. Navigate to their detail page and confirm the organization charts render as expected.
* Use unit search to find a unit with at least one parent. Navigate to its detail page and confirm it renders as expected.
